lib/realtime/block_waiver.ex

```diff
@@ -0,0 +1,113 @@
+defmodule Realtime.BlockWaiver do
+  @moduledoc """
+  A representation of known skipped work on a trip, entered by dispatch.
+  Essentially this summarizes a set of StopTimeUpdate datapoints.
+  """
+
+  alias Concentrate.StopTimeUpdate
+  alias Gtfs.{Block, Stop, Trip}
+  alias Realtime.Server
+
+  @type t :: %__MODULE__{
+          start_time: Util.Time.time_of_day(),
+          end_time: Util.Time.time_of_day(),
+          remark: String.t() | nil
+        }
+
+  @enforce_keys [
+    :start_time,
+    :end_time
+  ]
+
+  @derive Jason.Encoder
+
+  defstruct [
+    :start_time,
+    :end_time,
+    :remark
+  ]
+
+  @spec block_waivers_for_block(Block.t() | nil) :: [t()]
+  def block_waivers_for_block(nil), do: []
+
+  def block_waivers_for_block(block) do
+    block
+    |> Enum.flat_map(&trip_stop_time_waivers/1)
+    |> group_consecutive_sequences()
+    |> Enum.map(&from_trip_stop_time_waivers/1)
+  end
+
+  @type trip_stop_time_waiver ::
+          {Trip.id(), Stop.id(), Util.Time.time_of_day(), StopTimeUpdate.t() | nil}
```

Comment:
Instead of using this `trip_stop_time_waiver` as an intermediate type, I think you could make do with using block waivers directly.
* Transform each stop time update into a block waiver, where the start and end times are the time of the stop_time.
* Instead of the `acc` type being `[[trip_stop_time_waiver()]]`, it'd be `[t()]`.
* To merge them, see if the end time of one waiver and the start time of the next are close enough.

I think it'd cut down on some of the conversions, and it thematically makes sense: Each StopTimeUpdate is kind of its own block waiver that could hypothetically exist on its own, and we're doing the work to know that all those little block waivers are really one big block waiver.

Reply:
I see what you are saying. In my mind though I like the differentiation of intermediate data type for the intermediate steps, and then the final type is for the end results.

Reply:
Okay I see that. The `trip_id` and `stopid` fields of the `trip_stop_time_waiver` tuple are never used. Could they be removed?

Reply:
Okay I see that. The `trip_id` and `stopid` fields of the `trip_stop_time_waiver` tuple are never used. Could they be removed?